File: prototype/php/src/app/Actions/Auth/SendMagicLinkTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Auth;

use App\Domain\Auth\ActorType;
use App\Domain\Auth\MagicLinkToken;
use App\Models\MagicLink;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

final class SendMagicLinkTest extends TestCase
{
    use RefreshDatabase;

    public function test_it_stores_only_the_hash_of_the_token_it_delivers(): void
    {
        $url = $this->send('Artist@Example.com', ActorType::Seller);

        $token = basename(parse_url($url, PHP_URL_PATH));
        $link = MagicLink::sole();

        $this->assertSame(MagicLinkToken::hash($token), $link->token_hash);
        $this->assertStringNotContainsString($token, $link->token_hash);
    }

    public function test_it_normalizes_the_address_on_the_link(): void
    {
        $this->send('Artist@Example.com', ActorType::Seller);

        $this->assertSame('artist@example.com', MagicLink::sole()->email);
    }

    public function test_it_records_the_actor_the_link_signs_in(): void
    {
        $this->send('shopper@example.com', ActorType::Customer);

        $this->assertSame(ActorType::Customer, MagicLink::sole()->actor_type);
    }

    public function test_it_expires_the_link_after_the_configured_window(): void
    {
        config(['magic_links.expiry_minutes' => 15]);
        $this->freezeTime();

        $this->send('artist@example.com', ActorType::Seller);

        $this->assertSame(
            now()->addMinutes(15)->format('Y-m-d H:i:s'),
            MagicLink::sole()->expires_at->format('Y-m-d H:i:s'),
        );
    }

    public function test_it_carries_the_page_the_visitor_was_heading_for(): void
    {
        $this->send('shopper@example.com', ActorType::Customer, '/checkout');

        $this->assertSame('/checkout', MagicLink::sole()->redirect_to);
    }

    private function send(string $email, ActorType $actorType, ?string $redirectTo = null): string
    {
        $this->app->call(fn (SendMagicLink $send) => $send($email, $actorType, $redirectTo));

        return session('debug_magic_link');
    }
}

File: prototype/rails/src/test/actions/auth/send_magic_link_test.rb
require "test_helper"

module Auth
  class SendMagicLinkTest < ActiveSupport::TestCase
    class DeliveryRecorder
      attr_reader :email, :url

      def deliver(email:, url:)
        @email = email
        @url = url
      end
    end

    setup do
      @delivery = DeliveryRecorder.new
    end

    test "it issues a link for the address" do
      send_link(email: "artist@example.com")

      assert_equal "artist@example.com", MagicLink.sole.email
    end

    test "it normalizes the address before storing it" do
      send_link(email: "  Artist@Example.COM ")

      assert_equal "artist@example.com", MagicLink.sole.email
    end

    test "it stores only the digest of the token it hands out" do
      send_link

      token = @delivery.url.split("/").last

      assert_equal Domain::Auth::MagicLinkToken.digest(token), MagicLink.sole.token_digest
    end

    test "it delivers the verification url for the token" do
      send_link

      assert_match %r{\Ahttp://magic\.test/[0-9a-f]{64}\z}, @delivery.url
    end

    test "it delivers to the normalized address" do
      send_link(email: "Artist@Example.COM")

      assert_equal "artist@example.com", @delivery.email
    end

    test "it records which side of the marketplace asked" do
      send_link(actor_type: Domain::Auth::ActorType::CUSTOMER)

      assert_equal Domain::Auth::ActorType::CUSTOMER, MagicLink.sole.actor_type
    end

    test "it carries the destination the visitor was headed for" do
      send_link(redirect_to: "/orders/7/pay")

      assert_equal "/orders/7/pay", MagicLink.sole.redirect_to
    end

    test "it expires the link after the configured window" do
      freeze_time do
        send_link

        assert_equal Rails.configuration.x.magic_links.expiry_minutes.minutes.from_now, MagicLink.sole.expires_at
      end
    end

    test "it leaves the link unconsumed" do
      send_link

      assert_nil MagicLink.sole.consumed_at
    end

    test "two links for the same address carry different tokens" do
      send_link
      send_link

      assert_equal 2, MagicLink.distinct.count(:token_digest)
    end

    private

    def send_link(email: "artist@example.com", actor_type: Domain::Auth::ActorType::SELLER, redirect_to: nil)
      SendMagicLink
        .new(delivery: @delivery, link_url: ->(token) { "http://magic.test/#{token}" })
        .call(email: email, actor_type: actor_type, redirect_to: redirect_to)
    end
  end
end
